Use unaligned 32-bit and 64-bit compare in longest_match

nmoinvaz opened the pull request:
- Use unaligned 32-bit and 64-bit compare based on best match length when searching for matches.
- Move TRIGGER_LEVEL to match_tpl.h since it is only used in longest match.
- Use early return inside match loops instead of cont variable.
- Added back 2 byte check for platforms that don't supported unaligned access.

nmoinvaz committed:
Use unaligned 32-bit and 64-bit compare based on best match length when searching for matches.
Move TRIGGER_LEVEL to match_tpl.h since it is only used in longest match.
Use early return inside match loops instead of cont variable.
Added back two variable check for platforms that don't supported unaligned access.

nmoinvaz commented:
```
 ## HEAD
 
 Runs: 32
 Levels: 0-9
 Trimworst: 8

 Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.291/0.326/0.340                          211979883
 1      47.684%    1.115/1.222/1.262                          101065293
 2      35.540%    1.956/2.122/2.192                          75325919
 3      34.226%    2.311/2.499/2.555                          72541436
 4      32.953%    2.752/3.056/3.128                          69843427
 5      32.665%    3.032/3.262/3.341                          69233255
 6      32.515%    3.474/3.813/3.907                          68914325
 7      32.257%    4.752/5.133/5.287                          68367431
 8      32.172%    7.296/7.933/8.150                          68186831
 9      32.156% 10.326/10.972/11.180                          68154021

 avg1   41.218%                4.034
 avg2   45.798%                4.482
 tot                         968.156

## PR

 Runs: 32
 Levels: 0-9
 Trimworst: 8

  Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.290/0.317/0.342                          211979883
 1      47.684%    1.079/1.176/1.291                          101065293
 2      35.540%    1.850/2.020/2.191                          75326374
 3      34.203%    2.271/2.455/2.634                          72492718
 4      32.936%    2.762/2.980/3.198                          69806244
 5      32.665%    2.985/3.207/3.461                          69233255
 6      32.515%    3.487/3.741/3.988                          68914325
 7      32.257%    4.601/4.929/5.233                          68367431
 8      32.172%    7.095/7.577/8.090                          68186831
 9      32.156%  9.825/10.485/11.153                          68154021

 avg1   41.214%                3.889
 avg2   45.794%                4.321
 tot                         933.285
```

nmoinvaz commented on match_tpl.h:
```diff
@@ -35,23 +33,57 @@ int32_t LONGEST_MATCH(deflate_state *const s, Pos cur_match) {
+#define EARLY_EXIT_TRIGGER_LEVEL 5
+
+#define RETURN_BEST_LEN \
+    if (best_len <= s->lookahead) \
```
This if statement might not necessary since we already do such a check in `if (best_len >= nice_match) break`.

nmoinvaz committed:
Check for match length exceeding lookahead each time a new best match is found. This reduces some code complexity in GOTO_NEXT_CHAIN by removing the need for RETURN_BEST_LEN.

nmoinvaz commented:
New benchmarks. I ran it twice to see the margin of error, each time with 32 runs:

## HEAD 
```
 Runs: 32
 Levels: 0-9
 Trimworst: 8

 Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.290/0.308/0.320                          211979883
 1      47.684%    1.034/1.047/1.057                          101065293
 2      35.540%    1.812/1.844/1.870                          75325919
 3      34.226%    2.160/2.189/2.211                          72541436
 4      32.953%    2.619/2.640/2.666                          69843427
 5      32.665%    2.840/2.858/2.874                          69233255
 6      32.515%    3.323/3.349/3.368                          68914325
 7      32.257%    4.515/4.537/4.562                          68367431
 8      32.172%    6.977/7.007/7.035                          68186831
 9      32.156%    9.639/9.692/9.728                          68154021

 avg1   41.218%                3.547
 avg2   45.798%                3.941
 tot                         851.258

 Runs: 32
 Levels: 0-9
 Trimworst: 8

 Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.290/0.296/0.304                          211979883
 1      47.684%    1.024/1.050/1.103                          101065293
 2      35.540%    1.804/1.848/1.875                          75325919
 3      34.226%    2.149/2.190/2.238                          72541436
 4      32.953%    2.617/2.662/2.731                          69843427
 5      32.665%    2.811/2.865/2.934                          69233255
 6      32.515%    3.323/3.370/3.442                          68914325
 7      32.257%    4.512/4.562/4.678                          68367431
 8      32.172%    6.973/7.074/7.258                          68186831
 9      32.156%   9.651/9.755/10.007                          68154021

 avg1   41.218%                3.567
 avg2   45.798%                3.964
 tot                         856.146
```
## COMMIT 0
 ```
 Runs: 32
 Levels: 0-9
 Trimworst: 8

 Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.330/0.338/0.343                          211979883
 1      47.684%    1.030/1.045/1.061                          101065293
 2      35.540%    1.793/1.820/1.848                          75326374
 3      34.203%    2.181/2.225/2.245                          72492718
 4      32.936%    2.696/2.715/2.734                          69806244
 5      32.665%    2.906/2.927/2.938                          69233255
 6      32.515%    3.367/3.397/3.415                          68914325
 7      32.257%    4.476/4.493/4.514                          68367431
 8      32.172%    6.887/6.911/6.935                          68186831
 9      32.156%    9.520/9.541/9.579                          68154021

 avg1   41.214%                3.541
 avg2   45.794%                3.935
 tot                         849.883

 Runs: 32
 Levels: 0-9
 Trimworst: 8

 Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.289/0.294/0.299                          211979883
 1      47.684%    1.027/1.038/1.051                          101065293
 2      35.540%    1.785/1.807/1.840                          75326374
 3      34.203%    2.179/2.208/2.241                          72492718
 4      32.936%    2.675/2.700/2.714                          69806244
 5      32.665%    2.882/2.902/2.919                          69233255
 6      32.515%    3.351/3.380/3.401                          68914325
 7      32.257%    4.462/4.489/4.512                          68367431
 8      32.172%    6.854/6.891/6.918                          68186831
 9      32.156%    9.505/9.534/9.557                          68154021

 avg1   41.214%                3.524
 avg2   45.794%                3.916
 tot                         845.842
 ```
## COMMIT 1
```
 Runs: 32
 Levels: 0-9
 Trimworst: 8

 Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.288/0.297/0.304                          211979883
 1      47.684%    1.030/1.043/1.055                          101065293
 2      35.540%    1.758/1.779/1.794                          75326374
 3      34.203%    2.141/2.176/2.203                          72492718
 4      32.936%    2.631/2.655/2.681                          69806244
 5      32.665%    2.836/2.867/2.888                          69233255
 6      32.515%    3.315/3.334/3.348                          68914325
 7      32.257%    4.407/4.432/4.458                          68367431
 8      32.172%    6.812/6.842/6.863                          68186831
 9      32.156%    9.455/9.494/9.522                          68154021

 avg1   41.214%                3.492
 avg2   45.794%                3.880
 tot                         838.066

 Runs: 32
 Levels: 0-9
 Trimworst: 8

 Level   Comp   Comptime min/avg/max                          Compressed size
 0     100.015%    0.293/0.299/0.309                          211979883
 1      47.684%    1.024/1.036/1.085                          101065293
 2      35.540%    1.756/1.789/1.837                          75326374
 3      34.203%    2.136/2.177/2.210                          72492718
 4      32.936%    2.626/2.662/2.693                          69806244
 5      32.665%    2.833/2.859/2.886                          69233255
 6      32.515%    3.299/3.333/3.370                          68914325
 7      32.257%    4.389/4.435/4.501                          68367431
 8      32.172%    6.785/6.847/6.885                          68186831
 9      32.156%    9.442/9.490/9.546                          68154021

 avg1   41.214%                3.493
 avg2   45.794%                3.881
 tot                         838.258
```

Dead2 commented:
This looks very interesting. I see that the first commit changes the output size of compression levels 2-4 in my tests, do we know what specifically caused that change and that it is safe? In my tests it results in decreased size for all three levels, but I see that one of your tests ended up bigger. It looks like this change is only in the `level < TRIGGER_LEVEL` case.

My tests confirm that this results in a speedup, especially in the higher levels. In my tests level 9 is ~2.9% faster, and that is quite significant. For levels 1-2 the speed seems to be on par with before, with the change being well within the error margins.

Compiled code size increases quite a bit, but I am not too worried about that.

We should try to understand the nature of the output change and decide whether it is truly for the better, if it is then this looks very good indeed.

## Deflatebench config
 Runs: 70
 Levels: 1-9
 Trimworst: 40

Activated multiple generated file mode. Source: silesia-small.tar
Level 1: /tmp/deflatebench-1.tmp 270.1 MiB / 283,253,760 B
Level 2: /tmp/deflatebench-2.tmp 135.1 MiB / 141,626,880 B
Level 3: /tmp/deflatebench-3.tmp 105.1 MiB / 110,154,240 B
Level 4: /tmp/deflatebench-4.tmp 90.0 MiB / 94,417,920 B
Level 5: /tmp/deflatebench-5.tmp 90.0 MiB / 94,417,920 B
Level 6: /tmp/deflatebench-6.tmp 75.0 MiB / 78,681,600 B
Level 7: /tmp/deflatebench-7.tmp 60.0 MiB / 62,945,280 B
Level 8: /tmp/deflatebench-8.tmp 45.0 MiB / 47,208,960 B
Level 9: /tmp/deflatebench-9.tmp 45.0 MiB / 47,208,960 B


## Baseline 8832d7db7241194fa68509c96c092f3cf527ccce
```
   text    data     bss     dec     hex filename
  93335    1272      32   94639   171af libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.870%    1.863/1.880/1.891       1.181/1.215/1.234  163918126
 2      43.903%    1.825/1.854/1.863       0.614/0.631/0.640  62178636
 3      42.510%    1.700/1.724/1.734       0.460/0.476/0.483  46826710
 4      41.490%    1.681/1.699/1.707       0.387/0.399/0.407  39174432
 5      41.215%    1.825/1.849/1.857       0.372/0.396/0.405  38913920
 6      41.038%    1.834/1.845/1.851       0.323/0.332/0.337  32289128
 7      40.784%    1.998/2.014/2.021       0.253/0.261/0.269  25671760
 8      40.707%    1.935/1.949/1.955       0.185/0.194/0.199  19217358
 9      40.696%    2.154/2.164/2.170       0.185/0.195/0.201  19212387

 avg1   43.357%                1.887                   0.456
 tot                         509.364                 122.992
```

## PR #707 95eb3415c46aac9c79b6c5b2cb96b602a99f2ae7
```
   text    data     bss     dec     hex filename
  94871    1272      32   96175   177af libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.870%    1.852/1.876/1.888       1.176/1.212/1.231  163918126
 2      43.900%    1.845/1.857/1.868       0.615/0.630/0.638  62174655
 3      42.491%    1.751/1.767/1.776       0.463/0.477/0.486  46805381
 4      41.474%    1.712/1.726/1.735       0.378/0.397/0.405  39158901
 5      41.215%    1.855/1.870/1.880       0.380/0.395/0.406  38913920
 6      41.038%    1.814/1.841/1.851       0.315/0.331/0.338  32289128
 7      40.784%    1.954/1.966/1.972       0.253/0.263/0.269  25671760
 8      40.707%    1.879/1.894/1.901       0.182/0.195/0.201  19217358
 9      40.696%    2.090/2.101/2.106       0.178/0.195/0.201  19212387

 avg1   43.353%                1.878                   0.455
 tot                         506.955                 122.858
```

nmoinvaz commented:
With this PR it will check more bytes at the beginning and end of the match when the `best_len` allows. Since it does this check against more bytes it can more quickly identify if a match needs to be checked via `COMPARE256`. In _HEAD_ it would have had to check more potential matches, where as in this PR we can check less matches because we know more about the beginning and end of the match and whether or not they match our search string.

Dead2 commented on match_tpl.h:
```diff
@@ -89,50 +117,80 @@ int32_t LONGEST_MATCH(deflate_state *const s, Pos cur_match) {
-
-#if MIN_MATCH >= 2 && defined(UNALIGNED_OK)
-        len = COMPARE256(scan+2, match+2) + 2;
+        }
```
This } could go in front of the else if below, and the same with the next instance below where } can go in front of else, that way we can avoid the infamous "unbalanced else", and improving readability a tiny bit.

nmoinvaz replied:
Fixed.

Dead2 commented on match_tpl.h:
```diff
@@ -89,50 +117,80 @@ int32_t LONGEST_MATCH(deflate_state *const s, Pos cur_match) {
-                }
-                cont = 0;
+#ifdef UNALIGNED_OK
+        if (best_len < sizeof(uint32_t)) {
```
Here we check for smaller than sizeof(unit32_t)

nmoinvaz replied:
We do this check for best_len < 4 so we know to match using 2 byte match at beginning and end. This uses the `best_len-1` offset. 1 past byte plus 1 new byte.

Dead2 commented on match_tpl.h:
```diff
@@ -89,50 +117,80 @@ int32_t LONGEST_MATCH(deflate_state *const s, Pos cur_match) {
-            scan_end = *(bestcmp_t *)(scan+best_len-1);
+            offset = best_len-1;
+#ifdef UNALIGNED_OK
+            if (best_len >= sizeof(uint32_t)) {
```
..while here we check for larger than sizeof(unit32_t).

I have not read this code to the point of understanding yet, but this stood out a bit, two questions.
- Should these have been the same?
- If not, why is this case before the UNALIGNED64_OK case, and why is it doing both and not just one of them?
I am probably just missing something here, since it looks deliberate, but it is currently confusing me since there also are not any comments to detail the tricks. I just don't quite see why offset should be best_len-7.

mtl1979 replied:
bestlen-7 actually means bytes between bestlen-7 and bestlen, which is 8 bytes or 64 bits.
bestlen-3 actually means bytes between bestlen-3 and bestlen, which is 4 bytes or 32 bits.
bestlen-1 actually means bytes between bestlen-1 and bestlen, which is 2 bytes or 16 bits.

Dead2 replied:
Ah, that makes sense now :)

mtl1979 replied:
It took me a while to decipher, because normally programmers use 0-based index, but length starts with 1 when we are dealing something that isn't NULL-terminated as when start and end are the same memory location, it counts as length of 1.

nmoinvaz replied:
Yeah, we are positioning the read so it reads past 7 bytes of best_len and 1 extra byte for next best length. It can more quickly validate a potential match because it reads and compares more bytes of it toward the end and beginning.

nmoinvaz commented:
Rebased. Fixed syntax. Another potential commit added.

nmoinvaz commented:
## HEAD https://github.com/zlib-ng/zlib-ng/commit/193d8fd7dfb7927facab7a3034daa27ad5b9df1c
```
 Tool: minigzip-commit-0.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.761/1.775/1.784       0.689/0.700/0.704  75325919
 3      34.226%    2.101/2.110/2.119       0.675/0.683/0.687  72541436
 4      32.953%    2.542/2.555/2.564       0.657/0.664/0.667  69843427
 5      32.665%    2.752/2.762/2.769       0.655/0.659/0.663  69233255
 6      32.515%    3.237/3.251/3.263       0.649/0.657/0.662  68914325
 7      32.257%    4.404/4.420/4.431       0.647/0.656/0.660  68367431
 8      32.172%    6.837/6.855/6.870       0.646/0.656/0.660  68186831
 9      32.156%    9.469/9.492/9.511       0.651/0.657/0.660  68154021

 avg1   33.060%                4.152                   0.667
 tot                         996.598                 159.978
```
## COMMIT 1 https://github.com/zlib-ng/zlib-ng/pull/707/commits/fb42b016d4bade020db8bab474c2689f8ea8c046
```
 Tool: minigzip-commit-1.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.769/1.781/1.787       0.691/0.701/0.704  75326374
 3      34.203%    2.149/2.159/2.166       0.673/0.684/0.692  72492718
 4      32.936%    2.582/2.607/2.618       0.658/0.666/0.670  69806244
 5      32.665%    2.800/2.817/2.826       0.653/0.662/0.667  69233255
 6      32.515%    3.269/3.285/3.294       0.655/0.662/0.666  68914325
 7      32.257%    4.369/4.385/4.390       0.655/0.661/0.664  68367431
 8      32.172%    6.753/6.769/6.778       0.656/0.661/0.664  68186831
 9      32.156%    9.356/9.373/9.385       0.657/0.661/0.664  68154021

 avg1   33.055%                4.147                   0.670
 tot                         995.223                 160.749
```
## COMMIT 2 https://github.com/zlib-ng/zlib-ng/pull/707/commits/27aca49777f3eeeca2aa17c36fd3cb05bc7bf191
```
 Tool: minigzip-commit-2.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.731/1.741/1.748       0.691/0.698/0.703  75326374
 3      34.203%    2.117/2.129/2.139       0.673/0.682/0.687  72492718
 4      32.936%    2.577/2.591/2.598       0.657/0.665/0.670  69806244
 5      32.665%    2.786/2.795/2.803       0.654/0.662/0.666  69233255
 6      32.515%    3.245/3.261/3.271       0.653/0.661/0.664  68914325
 7      32.257%    4.319/4.333/4.340       0.650/0.660/0.664  68367431
 8      32.172%    6.691/6.716/6.723       0.655/0.660/0.664  68186831
 9      32.156%    9.307/9.324/9.334       0.657/0.661/0.664  68154021

 avg1   33.055%                4.111                   0.669
 tot                         986.685                 160.497
```
## COMMIT 3 https://github.com/zlib-ng/zlib-ng/pull/707/commits/f6ff56fa64edb00f9512c95574f1ed82e5a5b416
```
 Tool: minigzip-commit-3.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.706/1.718/1.726       0.689/0.696/0.701  75326425
 3      34.203%    2.085/2.100/2.107       0.672/0.679/0.683  72493099
 4      32.935%    2.552/2.568/2.579       0.648/0.660/0.664  69805892
 5      32.665%    2.757/2.773/2.782       0.651/0.658/0.661  69233255
 6      32.515%    3.224/3.237/3.246       0.649/0.655/0.660  68914325
 7      32.257%    4.294/4.306/4.315       0.645/0.653/0.657  68367431
 8      32.172%    6.652/6.673/6.685       0.647/0.654/0.657  68186831
 9      32.156%    9.251/9.270/9.283       0.648/0.653/0.658  68154021

 avg1   33.055%                4.081                   0.663
 tot                         979.362                 159.233

Dead2 commented:
Looks good to me.

nmoinvaz committed:
Precalculate match+offset pointer similar to fast-zlib.

nmoinvaz commented:
I have one additional commit pending.

nmoinvaz commented:
I will rerun the benchmarks tonight.

nmoinvaz commented:
## HEAD https://github.com/zlib-ng/zlib-ng/commit/193d8fd7dfb7927facab7a3034daa27ad5b9df1c
 ```
 Tool: minigzip-commit-0.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.750/1.762/1.771       0.688/0.694/0.699  75325919
 3      34.226%    2.084/2.095/2.105       0.672/0.679/0.683  72541436
 4      32.953%    2.517/2.534/2.547       0.652/0.659/0.664  69843427
 5      32.665%    2.721/2.742/2.753       0.651/0.657/0.661  69233255
 6      32.515%    3.213/3.227/3.237       0.645/0.652/0.657  68914325
 7      32.257%    4.376/4.391/4.400       0.643/0.649/0.652  68367431
 8      32.172%    6.793/6.806/6.819       0.644/0.651/0.655  68186831
 9      32.156%    9.395/9.417/9.429       0.645/0.651/0.654  68154021

 avg1   33.060%                4.122                   0.662
 tot                         989.251                 158.789
```
## COMMIT 1 https://github.com/zlib-ng/zlib-ng/pull/707/commits/fb42b016d4bade020db8bab474c2689f8ea8c046
```
 Tool: minigzip-commit-1.exe 
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.744/1.757/1.767       0.685/0.693/0.699  75326374
 3      34.203%    2.125/2.139/2.147       0.669/0.677/0.684  72492718
 4      32.936%    2.560/2.580/2.591       0.648/0.659/0.666  69806244
 5      32.665%    2.767/2.792/2.803       0.643/0.656/0.662  69233255
 6      32.515%    3.235/3.252/3.264       0.642/0.654/0.660  68914325
 7      32.257%    4.329/4.344/4.355       0.645/0.654/0.658  68367431
 8      32.172%    6.696/6.713/6.725       0.646/0.656/0.659  68186831
 9      32.156%    9.266/9.299/9.315       0.650/0.656/0.660  68154021

 avg1   33.055%                4.109                   0.663
 tot                         986.273                 159.110
```
## COMMIT 2 https://github.com/zlib-ng/zlib-ng/pull/707/commits/27aca49777f3eeeca2aa17c36fd3cb05bc7bf191
```
 Tool: minigzip-commit-2.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.708/1.726/1.735       0.685/0.692/0.700  75326374
 3      34.203%    2.091/2.111/2.124       0.667/0.675/0.683  72492718
 4      32.936%    2.546/2.563/2.576       0.644/0.657/0.664  69806244
 5      32.665%    2.752/2.764/2.778       0.644/0.655/0.663  69233255
 6      32.515%    3.217/3.233/3.243       0.641/0.654/0.661  68914325
 7      32.257%    4.269/4.293/4.302       0.646/0.654/0.657  68367431
 8      32.172%    6.637/6.658/6.668       0.651/0.655/0.658  68186831
 9      32.156%    9.217/9.242/9.254       0.649/0.655/0.658  68154021

 avg1   33.055%                4.074                   0.662
 tot                         977.696                 158.876
```
## COMMIT 3 https://github.com/zlib-ng/zlib-ng/pull/707/commits/379adef8964ab4753cb2424c0f2cc72a9695888a
```
 Tool: minigzip-commit-3.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.705/1.719/1.728       0.683/0.693/0.699  75326425
 3      34.203%    2.082/2.097/2.108       0.669/0.677/0.685  72493099
 4      32.935%    2.537/2.564/2.576       0.647/0.659/0.668  69805892
 5      32.665%    2.747/2.763/2.776       0.646/0.657/0.663  69233255
 6      32.515%    3.211/3.232/3.243       0.648/0.657/0.662  68914325
 7      32.257%    4.274/4.292/4.305       0.647/0.656/0.660  68367431
 8      32.172%    6.636/6.656/6.665       0.650/0.657/0.660  68186831
 9      32.156%    9.213/9.238/9.247       0.649/0.656/0.661  68154021

 avg1   33.055%                4.070                   0.664
 tot                         976.798                 159.338
```
## COMMIT 4 https://github.com/zlib-ng/zlib-ng/pull/707/commits/c813b4022246c118e444fa655c7ba7b14483c37a
```
 Tool: minigzip-commit-4.exe
 Runs: 70
 Levels: 2-9
 Trimworst: 40

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 2      35.540%    1.717/1.744/1.789       0.684/0.693/0.699  75326425
 3      34.203%    2.065/2.083/2.090       0.667/0.677/0.684  72493099
 4      32.935%    2.518/2.556/2.620       0.645/0.658/0.663  69805892
 5      32.665%    2.704/2.730/2.744       0.646/0.662/0.673  69233255
 6      32.515%    3.153/3.168/3.176       0.641/0.655/0.661  68914325
 7      32.257%    4.156/4.196/4.235       0.647/0.655/0.659  68367431
 8      32.172%    6.438/6.478/6.554       0.649/0.655/0.660  68186831
 9      32.156%    8.922/8.972/9.062       0.646/0.656/0.660  68154021

 avg1   33.055%                3.991                   0.664
 tot                         957.799                 159.325
```

Dead2 commented:
I feel like the "Compare only 2 bytes if best_len is less than 3 similar to fast-zlib." commit (379adef8964ab4753cb2424c0f2cc72a9695888a) should be separated out into its own PR and explained better. As-is it makes the code harder to understand and also changes the output. Better comments to explanain what is going on there is also needed.
Currently it is hard to grasp that best_len of length 3 has to fall through, while the first block comments talks about assuming the third byte is correct. So it is both confusing and changes output, I think that means it deserves a better look separately.

I see the new (last) commit (c813b4022246c118e444fa655c7ba7b14483c37a) also adds another register, increasing from the already high 4 to 5 requested static registers. That will likely cause the compiler to (even more often) disregard the hint for example on x86 platforms that only have the 8 GP registers in the first place.
It might be worth it to check what static registers are of the most benefit here and let the compiler handle those that do not, if the compiler disregards the hint due to register pressure, there is no telling what register(s) get dropped. And if it doesn't, then the compiler being forced to use 3 registers for all the other code will result in very sub-optimal code with frequent stores/loads to memory.
After this rewrite `window` is no longer part of the loop, so it no longer warrants living in a static register.
`len` is used a bit at the end, not sure that needs a reserved register though, the compiler is probably smart enough.
`scan` is not so often used either, but I am slightly more uncertain about this one.
Removing 2-3 of those would get us down to requesting 2-3 static registers, I suspect that might even improve performance on register-tight architectures, since the compiler is more likely to actually do what we asked.

Other than that, this looks very good, can't wait to get it merged.

nmoinvaz commented:
I have removed that extra commit as you said. I agree it does make things a bit more complicated and for minimal gain. I have removed some of the `ZLIB_REGISTER`s also and left it on for used the most variables in the for loops. I haven't done an investigation into which variables are best with `ZLIB_REGISTER`. My experience is some what limited there.

Dead2 commented:
## Baseline a19bbdeec892b1ff8210bd8c3b679cca77ad042d
```
   text    data     bss     dec     hex filename
  93151    1272      32   94455   170f7 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.869%    1.430/1.451/1.464       0.934/0.952/0.961  127491314
 2      43.903%    1.404/1.427/1.435       0.477/0.499/0.507  48360972
 3      42.509%    1.449/1.464/1.472       0.396/0.414/0.423  40136545
 4      41.492%    1.376/1.388/1.395       0.317/0.336/0.344  32646413
 5      41.214%    1.495/1.514/1.520       0.318/0.335/0.345  32427914
 6      41.038%    1.433/1.457/1.463       0.260/0.267/0.273  25831228
 7      40.784%    1.490/1.502/1.507       0.184/0.198/0.204  19253811
 8      40.706%    1.282/1.293/1.298       0.116/0.129/0.135  12811154
 9      40.695%    1.426/1.437/1.442       0.119/0.130/0.136  12807943

 avg1   43.357%                1.437                   0.362
 tot                         388.004                  97.793
```

## PR #707 87778ed7d37c966b58c2bb06c2630d6e4a1a3688
```
   text    data     bss     dec     hex filename
  94511    1272      32   95815   17647 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.869%    1.431/1.450/1.461       0.933/0.955/0.966  127491314
 2      43.900%    1.409/1.430/1.438       0.466/0.499/0.509  48357334
 3      42.491%    1.480/1.492/1.500       0.396/0.412/0.423  40118905
 4      41.473%    1.396/1.411/1.419       0.320/0.335/0.343  32631746
 5      41.214%    1.517/1.526/1.532       0.322/0.333/0.341  32427914
 6      41.038%    1.427/1.441/1.447       0.253/0.266/0.273  25831228
 7      40.784%    1.426/1.440/1.444       0.181/0.197/0.203  19253811
 8      40.706%    1.216/1.227/1.232       0.122/0.131/0.136  12811154
 9      40.695%    1.351/1.357/1.362       0.118/0.129/0.132  12807943

 avg1   43.352%                1.419                   0.362
 tot                         383.207                  97.751
```

Okay, so we still have an increase in compiled code size.
Average speedup is ~1.2%
Level 9 speedup is ~5.5%
Level 6 speedup is ~1.1%
With my testdata, levels 2-4 get a small reduction in compressed size, the rest appear unchanged.
